pkg/channel/consolidated/status/status_test.go

```diff
@@ -39,110 +39,169 @@ import (
-	}
-	port, err := strconv.Atoi(tsURL.Port())
-	if err != nil {
-		t.Fatalf("Failed to parse port %q: %v", tsURL.Port(), err)
+
+	lister := fakeProbeTargetLister{
+		target: getTargetLister(t, ts.URL),
 	}
-	hostname := tsURL.Hostname()
 
-	ready := make(chan *ReadyPair)
+	prober, ready := getProber(t, &lister)
+
+	done := make(chan struct{})
+	cancelled := prober.Start(done)
+	defer func() {
+		close(done)
+		<-cancelled
+	}()
+
+	assertEventuallyReady(t, prober, ch, sub, ready, &succeed, &probeRequests)
+}
+
+func TestProbeListerFail(t *testing.T) {
+	ch := getChannel(1)
+	sub := getSubscription()
 
+	ready := make(chan *ReadyPair)
+	defer close(ready)
 	prober := NewProber(
 		zaptest.NewLogger(t).Sugar(),
-		fakeProbeTargetLister{
-			PodIPs:  sets.NewString(hostname),
-			PodPort: strconv.Itoa(port),
-			URL:     tsURL,
-		},
+		notFoundLister{},
 		func(c v1beta1.KafkaChannel, s eventingduckv1.SubscriberSpec) {
 			ready <- &ReadyPair{
 				c,
 				s,
 			}
 		})
 
+	// If we can't list, this  must fail and return false
+	ok, err := prober.IsReady(context.Background(), *ch, *sub)
+	if err == nil {
+		t.Fatal("IsReady returned unexpected success")
+	}
+	if ok {
+		t.Fatal("IsReady() returned true")
+	}
+}
+
+func TestSucceedAfterRefreshPodProbing(t *testing.T) {
+	// We have a channel with three partitions
+	ch := getChannel(3)
+	sub := getSubscription()
+
+	// Dispatcher D1 will return only one ready partition
+	var subsD1 = map[string][]int{
+		string(sub.UID): {0},
+	}
+
+	// Dispatcher D2 will return the three ready partitions
+	var subsD2 = map[string][]int{
+		string(sub.UID): {0, 1, 2},
+	}
+
+	// The success handler for dispatcher D1, will return one partition only
+	successHandlerD1 := http.HandlerFunc(readyJSONHandler(t, subsD1))
+
+	// This is a latch channel that will lock the handler goroutine until we drain it
+	probeRequestsD1 := make(chan *http.Request)
+
+	handlerD1 := func(w http.ResponseWriter, r *http.Request) {
+		probeRequestsD1 <- r
+		successHandlerD1.ServeHTTP(w, r)
+	}
+
+	serverD1 := getDispatcherServer(handlerD1)
+	defer serverD1.Close()
+
+	// Probes only succeed if succeed is true
+	var succeed atomic.Bool
+
+	// The success handler for dispatcher D2, will return all three needed partitions
+	probeHandlerD2 := http.HandlerFunc(readyJSONHandler(t, subsD2))
+
+	// This is a latch channel that will lock the handler goroutine until we drain it
+	probeRequestsD2 := make(chan *http.Request)
+	handlerD2 := func(w http.ResponseWriter, r *http.Request) {
+		probeRequestsD2 <- r
+		if !succeed.Load() {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+
+		probeHandlerD2.ServeHTTP(w, r)
+	}
+
+	serverD2 := getDispatcherServer(handlerD2)
+	defer serverD2.Close()
+
+	// Initially, lister points to d1
+	lister := fakeProbeTargetLister{
+		target: getTargetLister(t, serverD1.URL),
+	}
+
+	prober, ready := getProber(t, &lister)
+
 	done := make(chan struct{})
 	cancelled := prober.Start(done)
```

Comment:
hm, just wondering out loud here, should we be `defer close(probeRequestsDX` elsewhere as well?

Reply:
It wouldn't hurt, but it's not needed IMHO. The reason those channels are closed is to shutdown the prober and free the polling goroutines in the prober and not leak those routines. While closing the probeRequests won't be necessary since we're closing the whole server anyway and that'll clean up their goroutines IIUC. And as you know, you don't need to close all the channels since they'll be garbage collected if unreachable.

If you feel I should do that, sure I can. Let me know.